File: packages/types/src/events.ts
import type { AssistantMessage } from "./agent";
import type { PermissionRequest, ToolCall, ToolCallPreview, ToolResult } from "./tool";
import type { TodoItem } from "./todo";

export interface AskQuestionRequest {
  requestId: string;
  question: string;
  /** Optional multiple-choice options; omit for a free-text question. */
  options?: string[];
  isMultiSelect?: boolean;
  /** When true, the user may skip the question entirely. */
  skippable?: boolean;
  /** Shared id grouping questions from a single askMany call. */
  batchId?: string;
}

export interface SubagentStartEvent {
  type: "subagentStart";
  subagentId: string;
  parentToolCallId: string;
  name: string;
  role: string;
  prompt: string;
  maxTurns: number;
}

export interface SubagentActivityEvent {
  type: "subagentActivity";
  subagentId: string;
  turnIndex: number;
  toolCallId: string;
  toolName: string;
  args?: Record<string, unknown>;
  status: "running" | "completed" | "error";
  error?: string;
}

export interface SubagentEndEvent {
  type: "subagentEnd";
  subagentId: string;
  status: "completed" | "aborted" | "error";
  summary?: string;
  error?: string;
  totalTurns: number;
}

export type SubagentEvent =
  | SubagentStartEvent
  | SubagentActivityEvent
  | SubagentEndEvent;

export interface SubagentActivityItem {
  turnIndex: number;
  toolCallId: string;
  toolName: string;
  args?: Record<string, unknown>;
  status: "running" | "completed" | "error";
  error?: string;
}

export interface SubagentInfo {
  subagentId: string;
  parentToolCallId: string;
  name: string;
  role: string;
  prompt: string;
  maxTurns: number;
  currentTurn: number;
  status: "running" | "completed" | "aborted" | "error";
  summary?: string;
  error?: string;
  activities: SubagentActivityItem[];
  createdAt?: number;
  updatedAt?: number;
}

export type AgentSessionEvent =
  | { type: "sessionStart" }
  | { type: "turnStart"; prompt: string }
  | { type: "modelStreamStart"; turnId: string }
  | { type: "modelStreamPart"; part: { text?: string; thinking?: string; toolCall?: ToolCallPreview } }
  | { type: "modelStreamEnd"; turnId: string; turn: AssistantMessage }
  | { type: "toolExecutionStart"; calls: ToolCall[] }
  | { type: "permissionRequest"; request: PermissionRequest }
  | { type: "askQuestion"; request: AskQuestionRequest }
  | { type: "toolExecutionResult"; result: ToolResult }
  | { type: "toolExecutionEnd"; results: ToolResult[] }
  | { type: "todoUpdate"; items: TodoItem[]; action: "created" | "updated" }
  | { type: "compaction"; summary: string; originalMessageCount: number }
  | { type: "turnEnd"; turnId: string }
  | { type: "sessionEnd" }
  | { type: "error"; error: { message: string; data?: unknown } }
  | SubagentStartEvent
  | SubagentActivityEvent
  | SubagentEndEvent
  /** Synthetic frame (re-attach streams only): run completed. */
  | { type: "done"; summary?: string }
  /** Synthetic frame (re-attach streams only): run was aborted. */
  | { type: "aborted"; reason?: string }
  /** Synthetic frame (re-attach streams only): precedes a replay batch so clients clear streaming buffers first. */
  | { type: "streamReset" };

export interface SseEventFrame {
  event: AgentSessionEvent["type"];
  data: AgentSessionEvent;
}

File: apps/desktop/crates/console-core/src/types/events.rs
use super::agent::{
    AskQuestionRequest, AssistantMessage, ImageAttachment, PermissionRequest, TodoItem, ToolCall,
    ToolCallPreview, ToolResult,
};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunPromptDto {
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attachments: Option<Vec<ImageAttachment>>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AnswerQuestionDto {
    pub request_id: String,
    pub answer: serde_json::Value,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApproveToolPermissionDto {
    pub request_id: String,
    pub allow: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum AgentSessionEvent {
    SessionStart,
    TurnStart {
        prompt: String,
    },
    ModelStreamStart {
        #[serde(rename = "turnId")]
        turn_id: String,
    },
    ModelStreamPart {
        part: ModelStreamPartPayload,
    },
    ModelStreamEnd {
        #[serde(rename = "turnId")]
        turn_id: String,
        #[serde(default)]
        turn: Option<AssistantMessage>,
    },
    ToolExecutionStart {
        calls: Vec<ToolCall>,
    },
    PermissionRequest {
        request: PermissionRequest,
    },
    AskQuestion {
        request: AskQuestionRequest,
    },
    ToolExecutionResult {
        result: ToolResult,
    },
    ToolExecutionEnd {
        results: Vec<ToolResult>,
    },
    TodoUpdate {
        items: Vec<TodoItem>,
        action: String,
    },
    Compaction {
        summary: String,
        original_message_count: usize,
    },
    TurnEnd {
        #[serde(rename = "turnId")]
        turn_id: String,
    },
    SubagentStart {
        #[serde(rename = "subagentId")]
        subagent_id: String,
        #[serde(rename = "parentToolCallId")]
        parent_tool_call_id: String,
        name: String,
        role: String,
        prompt: String,
        #[serde(rename = "maxTurns")]
        max_turns: usize,
    },
    SubagentActivity {
        #[serde(rename = "subagentId")]
        subagent_id: String,
        #[serde(rename = "turnIndex")]
        turn_index: usize,
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        #[serde(default)]
        args: Option<serde_json::Value>,
        status: String,
        #[serde(default)]
        error: Option<String>,
    },
    SubagentEnd {
        #[serde(rename = "subagentId")]
        subagent_id: String,
        status: String,
        #[serde(default)]
        summary: Option<String>,
        #[serde(default)]
        error: Option<String>,
        #[serde(rename = "totalTurns")]
        total_turns: usize,
    },
    SessionEnd,
    Error {
        error: ServerErrorPayload,
    },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelStreamPartPayload {
    pub text: Option<String>,
    pub thinking: Option<String>,
    pub tool_call: Option<ToolCallPreview>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServerErrorPayload {
    pub message: String,
    pub data: Option<serde_json::Value>,
}
